### DISCUSSION TITLE
run asnyc function in an endless loop

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#3065] [Q&A]
### Question

I made a client to Kepware throuth opcua,
the code works fine.how can I run this func in niceGui, and get the data 
```py
import sys
import asyncio
sys.path.insert(0, "..")
import logging
from asyncua import Client,ua

class SubHandler(object):
    """
    Subscription Handler. To receive events from server for a subscription
    """
    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)

    def event_notification(self, event):
        print("Python: New event", event)
async def main():
    url = "opc.tcp://localhost:49320"
    async with Client(url=url) as client:
        tag1 = client.get_node("ns=2;s=Channel1.Device1.Tag1")
        tag2 = client.get_node("ns=2;s=Channel1.Device1.Tag2")
        handler = SubHandler()
        sub = await client.create_subscription(500, handler)
        # handle1 = await sub.subscribe_data_change(tag1)
        handle2 = await sub.subscribe_data_change(tag2)
        while True:
            await asyncio.sleep(1)

        # await sub.unsubscribe(handle1)
        await sub.unsubscribe(handle2)
        await sub.delete()

if __name__ == "__main__":
    logging.basicConfig(level=logging.WARN)
    asyncio.run(main())
```

**Comment by falkoschindler** [comment]
Hi @v479038280,

> get data from opcserver

What is "opcserver"?

> I made a client to Kepware throuth opcua

What is "Kepware"? What is "opcua"?

> how can I run this func in niceGui, and get the data

What data?
What is the expected result/behavior?

I'll mark this question as "community question". Maybe someone from the community understands what you're trying to do and can help. But I kindly ask you to put a little more effort into your question. As we write in one of our [FAQs](https://github.com/zauberzeug/nicegui/wiki/FAQs): "Quite a number of people will read and think about your message. Make it worth their time."

**Reply by v479038280** [reply]
Imagine it as a kind of database that we connect to and retrieve data from. The current issue is that we have connected to it and obtained the data, but how do we transfer its data to NiceGUI? There are two ways to get data from it: one is to directly request data using tag1.read_value(), and the other is through subscription, where datachange_notification() is triggered when the data changes. More specifically, regarding how to put it into NiceGUI, the question is how to run asyncio.run(main()) in NiceGUI and pass the obtained data tag1, tag2 to NiceGUI for display or other operations.

**Comment by v479038280** [comment]
>What is "opcserver"?

An OPC Server (OLE for Process Control Server) is a software component that adheres to the OPC (OLE for Process Control) standards, facilitating the exchange of data between devices or systems within industrial automation and control systems. OPC serves as an open, platform-independent specification designed to enhance interoperability among equipment and systems from various manufacturers.

>What is "Kepware"? What is "opcua"?

 [here](https://opcfoundation.org/about/opc-technologies/opc-ua/) is an article about opcua, 。
Kepware is a software that integrates a multitude of industrial device drivers, which allows it to read data from the devices and then act as an OPC (Open Platform Communications) server. Once we connect to it, we can access the data from various devices without having to connect to each device individually or create data connection drivers for each one. 

Here's a more detailed explanation in English:

1. **Kepware as a Software**: Kepware is a software application designed to serve as a communication conduit between industrial devices and other systems.

2. **Integration of Device Drivers**: It incorporates a wide range of device drivers, enabling it to interface with numerous industrial devices.

3. **Data Collection**: Kepware is capable of reading data from these devices, aggregating it in a single location.

4. **OPC Server Role**: After collecting the data, Kepware can function as an OPC server, which is a software component that adheres to the OPC (Open Platform Communications) standards for data exchange.

5. **Simplified Data Access**: By connecting to the OPC server provided by Kepware, users can retrieve data from multiple devices without the need to establish individual connections to each device.

6. **Elimination of Individual Device Drivers**: There is no need to develop or implement specific data connection drivers for each device when using Kepware, as it already provides the necessary interfaces.

7. **Efficiency and Convenience**: This approach streamlines the process of data collection and simplifies the integration of industrial devices into a unified system or platform.

>What data? What is the expected result/behavior?

In the example code above,the data is the variable tag1 and tag2, the expected result is display the data in nicegui,for example ui.label(f'{tag1}),  and ui.label(f'{tag2})

**Reply by v479038280** [reply]
**Actually,I think there is no need to think What's opcua, just think it as some kind of dataserver, we connect to it and get the  data,in the code above  tag1 and tag2, they are just  variables that  store numbers.  And I have get the data.**

@falkoschindler  Thank you for your reply,I think  here is the key about the question
**So my actual question is how to run an async/await function, the main() function above, within NiceGUI, and keep it running in the background, displaying the read variables tag1 and tag2 in ui.label(f'{tag1}') and ui.label(f'{tag2}').  
There is absolutely no need to focus on OPC; instead, we just need to concentrate on how Python and NiceGUI handle this type of function.**

I made some change to the code above ,to make it easy to understand.  this works fine. but how to make it run in a endless loop
```
async def clientToKepwareV5():
    url = "opc.tcp://localhost:49320"
    async with Client(url=url) as client:
        tag2 = client.get_node("ns=2;s=Channel1.Device1.Tag2")
        ui.notify(f"tag2 is: {tag2} with value {await tag2.read_value()} ")

ui.button('start async task', on_click=clientToKepwareV5)

ui.run()
```

**Comment by v479038280** [comment]
I close this question ,and made a new on #3067,That would be clearer and less confusing.
